File: Leetcode567.java
public class Leetcode567 {
    public boolean checkInclusion(String s1, String s2) {
        if (s1==null || s2==null) {return false;}

        int L1 = s1.length();
        int L2 = s2.length();

        if(L1>L2) {return false;}

        int[] charCount = new int[26];

        for(char c1: s1.toCharArray()){
            charCount[c1 - 'a'] += 1;
        }

        int left=0, right=0;

        // Move the window
        while(left <= L2-L1){

            // Extend the window
            // window length < L1 and there is left character for extension
            while(right-left<L1 && charCount[s2.charAt(right)-'a']>=1){
                charCount[s2.charAt(right) - 'a'] -= 1;
                right += 1;
            }

            if(right-left == L1) {return true;}
            // if window length != L1 or no character for extension
            // move the window
            else{
                charCount[s2.charAt(left) - 'a'] += 1;
                left += 1;
            }
        }
        return false;
    }
}

// Count the alphabet in s1 and use them in s2.
// if window length is smaller then s1.length and there is left alphabet, extend the window
// else, move the window for 1 index.

File: Python/Leetcode567.py
class Leetcode567:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if not s1 or not s2: return False

        l1, l2 = len(s1), len(s2)
        if l1>l2: return False

        charCount = [0] * 26

        for c1 in s1:
            charCount[ord(c1) - ord('a')] += 1

        left = right = 0

        while left <= l2 - l1:
            # 오른쪽으로 늘일 수 있으면 늘린다
            while (right - left < l1) and charCount[ord(s2[right]) - ord('a')] >= 1:
                charCount[ord(s2[right]) - ord('a')] -= 1
                right += 1
            
            if right - left == l1:
                return True
            # 더 이상 늘일 수 없어서 빠져나온 경우(조건만족 못 한 경우) -> 왼쪽 한 칸 옮겨서 반복
            else:
                charCount[ord(s2[left]) - ord('a')] += 1
                left += 1

        return False




        




        




# def checkInclusion(self, s1: str, s2: str) -> bool:
#     l1, l2 = len(s1), len(s2)
#     count1 = Counter(s1)

#     for i in range(l2 - l1 + 1):
#         count2 = Counter(s2[i:i+l1])
#         if count2 == count1:
#             return True

#     return False